src/main/java/org/jboss/modules/JDKSpecific.java

```diff
@@ -185,6 +189,19 @@ static void addInternalPackages(final List<String> list) {
         list.add("jdk.internal.reflect.");
     }
 
+    static <T> Iterable<T> findServices(final Class<T> serviceType, final Predicate<Class<?>> filter, final ClassLoader classLoader) {
+        final Iterator<T> i = ServiceLoader.load(serviceType, classLoader).iterator();
+        final List<T> providers = new LinkedList<>();
+        T provider;
+        while (i.hasNext()) {
+            provider = i.next();
+            if (provider != null && filter.test(provider.getClass())) {
+                providers.add(provider);
+            }
+        }
+        return providers;
+    }
+
```

Comment:
Other than `LinkedList` being a class that should *never* be used, I'd change to something like this:

```suggestion
        ServiceLoader<T> sl = ServiceLoader.load(serviceType, classLoader);
        return new Iterable<T>() {
            public Iterator<T> iterator() {
                final Iterator<T> delegate = sl.iterator();
                return new Iterator<T>() {
                    T next = null;
                    public boolean hasNext() {
                        while (next == null) {
                            if (! delegate.hasNext()) {
                                return false;
                            }
                            next = delegate.next();
                            if (filter.test(next.getClass())) {
                                this.next = next;
                                return true;
                            }
                        }
                        return true;
                    }
                    
                    public T next() {
                        if (! hasNext()) throw new NoSuchElementException();
                        T next = this.next;
                        this.next = null;
                        return next;
                    }
                }
            }
        }
    }

```

Reply:
Thanks for proposal, that's really better (because of lazyness)